pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,410 @@
+	})
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK, podGVK}))
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: nsedConfigInstanceOne}: nil,
+		{Gvk: podGVK, Key: nsedPodInstanceOne}:          nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+
+	fi.setFailures("ConfigMapList", 5)
+
+	// this call should schedule a cache wipe and a replay for the configMapGVK
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+
+	expected2 := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: nsedConfigInstanceOne}: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected2), eventuallyTimeout, eventuallyTicker)
+}
+
+// TestCacheManager_concurrent makes sure that we can add and remove multiple sources
+// from separate go routines and changes to the underlying cache are reflected.
+func TestCacheManager_concurrent(t *testing.T) {
+	r := rand.New(rand.NewSource(12345)) // #nosec G404: Using weak random number generator for determinism between calls
+
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+	testResources, ctx := makeTestResources(t, mgr, wm, c)
+
+	cacheManager := testResources.CacheManager
+	dataStore := testResources.CFDataClient
+	agg := testResources.GVKAgreggator
+
+	// Create configMaps to test for
+	cm := unstructuredFor(configMapGVK, configInstancOne)
+	require.NoError(t, c.Create(ctx, cm), fmt.Sprintf("creating ConfigMap %s", configInstancOne))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, cm), fmt.Sprintf("deleting resource %s", configInstancOne))
+	})
+
+	cm2 := unstructuredFor(configMapGVK, configInstancTwo)
+	require.NoError(t, c.Create(ctx, cm2), fmt.Sprintf("creating ConfigMap %s", configInstancTwo))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, cm2), fmt.Sprintf("deleting resource %s", configInstancTwo))
+	})
+
+	pod := unstructuredFor(podGVK, podInstanceOne)
+	require.NoError(t, c.Create(ctx, pod), fmt.Sprintf("creating Pod %s", podInstanceOne))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, pod), fmt.Sprintf("deleting resource %s", podInstanceOne))
+	})
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	syncSourceTwo := aggregator.Key{Source: "source_b", ID: "ID_b"}
+
+	wg := &sync.WaitGroup{}
+
+	// simulate a churn-y concurrent access by swapping the gvks for the sync sources repeatedly
+	// and removing sync sources, all from different go routines.
+	for i := 1; i < 100; i++ {
+		wg.Add(3)
+		go func() {
+			defer wg.Done()
+
+			// add some jitter
+			time.Sleep(time.Duration(r.Intn(1000)) * time.Millisecond)
```

Comment:
Can we move the 1000 to a constant? Also lets drop it to 500 just to lower the maximum test runtime.